mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala

```diff
@@ -100,19 +100,20 @@ class HashingTF @Since("1.4.0") (@Since("1.4.0") override val uid: String)
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema)
+    val localNumFeatures = $(numFeatures)
+    val localBinary = $(binary)
+
     val hashUDF = udf { terms: Seq[_] =>
-      val numOfFeatures = $(numFeatures)
-      val isBinary = $(binary)
       val termFrequencies = mutable.HashMap.empty[Int, Double].withDefaultValue(0.0)
       terms.foreach { term =>
-        val i = indexOf(term)
-        if (isBinary) {
+        val i = Utils.nonNegativeMod(hashFunc(term), localNumFeatures)
```

Comment:
OK, yeah I think we had discussed whether it's better to reuse indexOf or call this directly with a local var. I'm OK with either.

Reply:
Since we now have ```indexOf```, I guess it might be slight better to reuse this method.

Reply:
Yeah I'm neutral on it. As a matter of software design, reusing `indexOf` makes sense. It may be more performant to inline it. Is there evidence it makes a performance difference?